SwiftPrivilegedHelperApplication/Scripts/CodeSignUpdate.swift

```diff
@@ -0,0 +1,402 @@
+func environmentVarNotSet(_ variable: String, addHelp: Bool = false) -> Never
+{
+    let message = "\(variable) environment variable not set\n\(helpStr)"
+        + (addHelp ? "\n\(helpStr)" : "")
+    
+    emitError(message)
+}
+
+// -------------------------------------
+func getBundleIDFrom(_ environmentVariable: String) -> String
+{
+    guard let bundleID = Environment[environmentVariable] else {
+        environmentVarNotSet(environmentVariable, addHelp: true)
+    }
+    
+    let bundleIDChars = alphaNumericChars + ["."]
+    
+    guard bundleID.count > 1 else {
+        emitError("Bundle ID is empty\n\(bundleIDFormatHelp)")
+    }
+    
+    guard bundleID.reduce(true, { $0 && bundleIDChars.contains($1) }) else
+    {
+        emitError(
+            "Bundle ID contains illegal characters: \(bundleID)\n"
+            + "\(bundleIDFormatHelp)"
+        )
+    }
+    
+    guard bundleID.first! != "." && bundleID.last! != "." else
+    {
+        emitError(
+            "Bundle ID must not start or end with period: \(bundleID)\n"
+            + "\(bundleIDFormatHelp)"
+        )
+    }
+    
+    return bundleID
+}
+
+
+let bundleIdentifierApplication = getBundleIDFrom("MAIN_BUNDLE_ID")
+let bundleIdentifierHelper = getBundleIDFrom("HELPER_BUNDLE_ID")
+
+guard let infoPListFile = Environment["INFOPLIST_FILE"] else {
+    environmentVarNotSet("INFOPLIST_FILE")
+}
+
+var infoPlist: [String: AnyObject] =
+{
+    let pListURL = URL(fileURLWithPath: infoPListFile)
+    guard let pListDict = NSDictionary(contentsOf: pListURL)
+        as? Dictionary<String, AnyObject>
+    else {
+        emitError("Unable to read plist data from \(infoPListFile)")
+    }
+    return pListDict
+}()
+
+let target: String = infoPlist["NSPrincipalClass"] as? String == "NSApplication"
+    ? "application"
+    : "helper"
+
+// -------------------------------------
+func isValidDeveloperID<S: StringProtocol>(_ s: S) -> Bool
+{
+    guard s.count == 10 else { return false }
+    return s.reduce(true) { $0 && devIDChars.contains($1) }
+}
+
+// -------------------------------------
+func isValidDeveloperCN(_ s: String, withPrefix prefix: String) -> Bool
+{
+    // +4 accounts for 2 spaces, plus open and close parentheses
+    guard s.hasPrefix(prefix), s.count > prefix.count + 4 else { return false }
+    
+    guard let openParen = s.lastIndex(of: "(") else { return false }
+    guard let closeParen = s.lastIndex(of: ")") else { return false }
+    
+    guard s.distance(from: closeParen, to: s.endIndex) == 1 else {
+        return false
+    }
+    
+    guard s.distance(from: openParen, to: closeParen) == 11 else {
+        return false
+    }
+    
+    return isValidDeveloperID(s[s.index(after: openParen)..<closeParen])
+}
+
+// -------------------------------------
+func isValidDeveloperCN(_ s: String) -> Bool
+{
+    let appleDeveloper = "Apple Development:"
+    let macDeveloper = "Mac Developer:"
+    let customDeveloper = "Developer ID Application:"
```

Comment:
When signing with a Developer ID certificate, you need to match the logic and structure of the designated requirements that are generated when `action == install`.  However, `isValidDeveloperCN` only gets called by way of `appendMacDeveloper` - which isn't called for `action == install`.  I'm not sure that there is benefit in supporting the Developer ID certificate when `action == build`.

(This is just a suggestion and some feedback.  The failure caused by trying to sign with a developer ID cert under `action == build` is what helped me to realize that my build scripts were not correct for my development versus deployment profiles...with this change, I would not have had that error to guide me in the correct direction).

Reply:
Another alternative (which I think might be even more helpful) is to key off of the Developer ID in the signing certificate, and if it is a developer ID cert, calculate like you currently do for `action == install`, and if it's non-Developer ID, calculate like you currently do for `action == build`.  That would match the correct designated requirements for _any_ certificate (and then you wouldn't key off the `ACTION` environment variable anymore).

As it stands, I always to my builds with the `build` target, so I need to `export ACTION=install` when I'm signing with a developer ID cert (and `export ACTION=build` when I'm signing with my dev certificates).  But - keying off the cert might be nicer than keying off of the `ACTION` environment variable.  Again - just suggestions.